Add datalinks support to PieChart v2

oscarkilhed opened the pull request:
**What this PR does / why we need it**:
This adds missing data link support for the PieChart

**Special notes for your reviewer**:
Not sure I like how the SVG component is responsible for the context menu, ideally this would injected as a wrapper somehow. But did not find a good way to do that.

oscarkilhed committed:
Remove default value for multiSelect

oscarkilhed committed:
Add data links support for PieChart v2

oscarkilhed committed:
Temporarily fix story

oscarkilhed committed:
Merge branch 'master' into oscark/add-datalinks-support-for-piechart

dprokop commented on packages/grafana-ui/src/components/DataLinks/DataLinksContextMenu.tsx:
```diff
@@ -10,7 +10,7 @@ interface DataLinksContextMenuProps {
 
 export interface DataLinksContextMenuApi {
-  openMenu?: React.MouseEventHandler<HTMLElement>;
+  openMenu?: React.MouseEventHandler<HTMLElement | SVGElement>;
```
Wonder if we can use HTMLOrSVGElement?

dprokop commented on public/app/plugins/panel/piechart/PieChartPanel.tsx:
```diff
@@ -23,20 +23,20 @@ export const PieChartPanel: React.FC<Props> = ({
       width={width}
       height={height}
-      values={values}
+      fieldDisplayValues={fieldDisplayValues}
```
Not directly related to data links, but maybe we could follow one direction when it comes to panel -> visualization data flow? Instead of passing fieldDisplayValues I would expect PieChart to accept dataFrame(s) and figure out the field display value on it's own. Then panel's responsibility os only to manipulate the options and the visualiation handles iz specific operations?

WDYT @torkelo ?

torkelo replied:
yea, that sounds like a good change, should make it easier to re-use in other contexts

torkelo replied:
added that to https://github.com/grafana/grafana/issues/31116

dprokop commented on packages/grafana-ui/src/components/PieChart/PieChart.tsx:
```diff
@@ -187,31 +224,15 @@ export const PieChartSvg: FC<SvgProps> = ({
-                    )}
-                  </g>
-                );
+                if (arc.data.hasLinks && arc.data.getLinks) {
+                  return (
+                    <DataLinksContextMenu key={arc.index} links={arc.data.getLinks}>
+                      {(api) => getPieSlice(arc, pie, api.openMenu)}
+                    </DataLinksContextMenu>
+                  );
```
In a situation when there is only a single data link I would expect no context menu and only the arc being clickable. Ref: https://github.com/grafana/grafana/issues/26805

torkelo replied:
@dprokop think the change to https://github.com/grafana/grafana/pull/31692 I think this will be a bit tricky to use here as DataLinksContextMenu renders anchor , and I do not think an anchor is allowed inside an svg.

dprokop replied:
Shit, I think they are allowed but require `xlink:href` attribute instead of `href`

dprokop commented on packages/grafana-ui/src/components/PieChart/PieChart.tsx:
```diff
@@ -153,18 +157,51 @@ export const PieChartSvg: FC<SvgProps> = ({
+  const total = fieldDisplayValues.reduce((acc, item) => item.display.numeric + acc, 0);
   const layout = getPieLayout(width, height, pieType);
 
+  const getPieSlice = (
```
This feels like a react component to me. What about having a PieSlice component that would handle data links in it rather than this render function in a component? From PieChartSvg render POV it would look like this:
```return pie.arcs.map(arc => <PieSlice .../>)```

torkelo reviewed (commented):
Awesome job getting this working! data links is a tricky (and a bit messy area). 

Noticed one things, tooltips are not showing in panel edit mode (probably z-index issue). 

Agree with some of dominiks comments. Would be great if there was only one link the slice would could be clicked directly. Maybe this can handled directly by DataLinksContextMenu ? 

We need to somehow make the fact that there is a data link there visible in the tooltip, might need to change the data we pass to the tooltip from DisplayValue to the full FieldDisplay so we can access hasLinks there (might need to call getLinks as well to show the name of the first link if it's only one).

torkelo commented:
@oscarkilhed unrelated to this issue found a bug when you set to slices to the same color:

```
Warning: Encountered two children with the same key, `#5794F2`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.
    at g
```

oscarkilhed commented:
> @oscarkilhed unrelated to this issue found a bug when you set to slices to the same color:
> 
> ```
> Warning: Encountered two children with the same key, `#5794F2`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.
>     at g
> ```

Yeah I noticed this as well, should be fairly easy to solve by deduplicating the colors before creating the RadialGradient elements.

oscarkilhed committed:
Refactor PieSlice, deduplicate colors

oscarkilhed committed:
Merge branch 'master' into oscark/add-datalinks-support-for-piechart

oscarkilhed committed:
Add field config to context menu of pie chart

oscarkilhed commented:
Tooltips still need some work

oscarkilhed committed:
Add custom key to legend

This is a bit hacky. When there are multiple DisplayValues with the same
title the react keys collide and LegendListItems are not cleared
correctly.

oscarkilhed committed:
Forgot to add the interface

oscarkilhed committed:
Fix missing tooltip in edit mode

oscarkilhed commented:
I suggest we defer showing that there are existing links on the pie slice tooltip until we change the tooltip to use our own.

torkelo reviewed (approved):
Think this looks great!
